# EDA (Exploratory Data Analysis)
***
## We have here a complex dataset (3 .csv files) which we have to deal with it carefully and extract only the data which will help us build our dashboard application .
***

In [1]:
#import the dependencies
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

#options
%matplotlib inline
pd.set_option('display.max_columns',None)

In [2]:
main_df = pd.read_csv('Stats.csv')
countries_df = pd.read_csv('StatsCountry.csv')
indicators_df = pd.read_csv('StatsSeries.csv')

main_df.head()

Country Name Country Code  \
0   Arab World          ARB   
1   Arab World          ARB   
2   Arab World          ARB   
3   Arab World          ARB   
4   Arab World          ARB   

                                      Indicator Name  Indicator Code  \
0  Adjusted net enrolment rate, lower secondary, ...      UIS.NERA.2   
1  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.F   
2  Adjusted net enrolment rate, lower secondary, ...  UIS.NERA.2.GPI   
3  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.M   
4  Adjusted net enrolment rate, primary, both sex...     SE.PRM.TENR   

        1970       1971       1972       1973       1974      1975       1976  \
0        NaN        NaN        NaN        NaN        NaN       NaN        NaN   
1        NaN        NaN        NaN        NaN        NaN       NaN        NaN   
2        NaN        NaN        NaN        NaN        NaN       NaN        NaN   
3        NaN        NaN        NaN        NaN        NaN       NaN        NaN   
4  54.822121  54.894138  56.209438  57.267109  57.991138  59.36554  60.999962   

       1977      1978       1979       1980       1981       1982       1983  \
0       NaN       NaN        NaN        NaN        NaN        NaN        NaN   
1       NaN       NaN        NaN        NaN        NaN        NaN        NaN   
2       NaN       NaN        NaN        NaN        NaN        NaN        NaN   
3       NaN       NaN        NaN        NaN        NaN        NaN        NaN   
4  61.92268  62.69342  64.383186  65.617767  66.085152  66.608139  67.290451   

        1984       1985       1986      1987       1988       1989       1990  \
0        NaN        NaN        NaN       NaN        NaN        NaN        NaN   
1        NaN        NaN        NaN       NaN        NaN        NaN        NaN   
2        NaN        NaN        NaN       NaN        NaN        NaN        NaN   
3        NaN        NaN        NaN       NaN        NaN        NaN        NaN   
4  68.510094  69.033211  69.944908  71.04187  71.693779  71.699097  71.995819   

        1991       1992       1993       1994      1995       1996       1997  \
0        NaN        NaN        NaN        NaN       NaN        NaN        NaN   
1        NaN        NaN        NaN        NaN       NaN        NaN        NaN   
2        NaN        NaN        NaN        NaN       NaN        NaN        NaN   
3        NaN        NaN        NaN        NaN       NaN        NaN        NaN   
4  72.602837  70.032722  70.464821  72.645683  71.81176  73.903511  74.425201   

        1998       1999       2000       2001       2002       2003  \
0        NaN        NaN        NaN        NaN        NaN        NaN   
1        NaN        NaN        NaN        NaN        NaN        NaN   
2        NaN        NaN        NaN        NaN        NaN        NaN   
3        NaN        NaN        NaN        NaN        NaN        NaN   
4  75.110817  76.254318  77.245682  78.800522  80.051399  80.805389   

        2004       2005       2006       2007       2008       2009  \
0        NaN        NaN        NaN        NaN        NaN        NaN   
1        NaN        NaN        NaN        NaN        NaN        NaN   
2        NaN        NaN        NaN        NaN        NaN        NaN   
3        NaN        NaN        NaN        NaN        NaN        NaN   
4  81.607063  82.489487  82.685509  83.280342  84.011871  84.195961   

        2010      2011       2012      2013       2014  2015  2016  2017  \
0        NaN       NaN        NaN       NaN        NaN   NaN   NaN   NaN   
1        NaN       NaN        NaN       NaN        NaN   NaN   NaN   NaN   
2        NaN       NaN        NaN       NaN        NaN   NaN   NaN   NaN   
3        NaN       NaN        NaN       NaN        NaN   NaN   NaN   NaN   
4  85.211998  85.24514  86.101669  85.51194  85.320152   NaN   NaN   NaN   

   2020  2025  2030  2035  2040  2045  2050  2055  2060  2065  2070  2075  \
0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

In [3]:
countries_df.head()

Country Code   Short Name   Table Name                     Long Name  \
0          ABW        Aruba        Aruba                         Aruba   
1          AFG  Afghanistan  Afghanistan  Islamic State of Afghanistan   
2          AGO       Angola       Angola   People's Republic of Angola   
3          ALB      Albania      Albania           Republic of Albania   
4          AND      Andorra      Andorra       Principality of Andorra   

  2-alpha code   Currency Unit  \
0           AW   Aruban florin   
1           AF  Afghan afghani   
2           AO  Angolan kwanza   
3           AL    Albanian lek   
4           AD            Euro   

                                       Special Notes  \
0  SNA data for 2000-2011 are updated from offici...   
1  Fiscal year end: March 20; reporting period fo...   
2  April 2013 database update: Based on IMF data,...   
3                                                NaN   
4                                                NaN   

                      Region          Income Group WB-2 code  \
0  Latin America & Caribbean  High income: nonOECD        AW   
1                 South Asia            Low income        AF   
2         Sub-Saharan Africa   Upper middle income        AO   
3      Europe & Central Asia   Upper middle income        AL   
4      Europe & Central Asia  High income: nonOECD        AD   

                         National accounts base year  \
0                                               2000   
1                                            2002/03   
2                                               2002   
3  Original chained constant price data are resca...   
4                                               1990   

   National accounts reference year                   SNA price valuation  \
0                               NaN     Value added at basic prices (VAB)   
1                               NaN     Value added at basic prices (VAB)   
2                               NaN  Value added at producer prices (VAP)   
3                            1996.0     Value added at basic prices (VAB)   
4                               NaN                                   NaN   

  Lending category Other groups  \
0              NaN          NaN   
1              IDA         HIPC   
2             IBRD          NaN   
3             IBRD          NaN   
4              NaN          NaN   

                         System of National Accounts  \
0  Country uses the 1993 System of National Accou...   
1  Country uses the 1993 System of National Accou...   
2  Country uses the 1993 System of National Accou...   
3  Country uses the 1993 System of National Accou...   
4  Country uses the 1968 System of National Accou...   

  Alternative conversion factor PPP survey year  \
0                           NaN             NaN   
1                           NaN             NaN   
2                       1991–96            2005   
3                           NaN         Rolling   
4                           NaN             NaN   

              Balance of Payments Manual in use  \
0  IMF Balance of Payments Manual, 6th edition.   
1                                           NaN   
2  IMF Balance of Payments Manual, 6th edition.   
3  IMF Balance of Payments Manual, 6th edition.   
4                                           NaN   

  External debt Reporting status       System of trade  \
0                            NaN  Special trade system   
1                         Actual  General trade system   
2                         Actual  Special trade system   
3                         Actual  General trade system   
4                            NaN  Special trade system   

     Government Accounting concept           IMF data dissemination standard  \
0                              NaN                                       NaN   
1  Consolidated central government  General Data Dissemination System (GDDS)   
2     Budgetary central government  General Data Dissemination System (GDDS)   

In [4]:
indicators_df.head()

Series Code       Topic  \
0  BAR.NOED.1519.FE.ZS  Attainment   
1     BAR.NOED.1519.ZS  Attainment   
2  BAR.NOED.15UP.FE.ZS  Attainment   
3     BAR.NOED.15UP.ZS  Attainment   
4  BAR.NOED.2024.FE.ZS  Attainment   

                                      Indicator Name  \
0  Barro-Lee: Percentage of female population age...   
1  Barro-Lee: Percentage of population age 15-19 ...   
2  Barro-Lee: Percentage of female population age...   
3  Barro-Lee: Percentage of population age 15+ wi...   
4  Barro-Lee: Percentage of female population age...   

                                    Short definition  \
0  Percentage of female population age 15-19 with...   
1  Percentage of population age 15-19 with no edu...   
2  Percentage of female population age 15+ with n...   
3  Percentage of population age 15+ with no educa...   
4  Percentage of female population age 20-24 with...   

                                     Long definition  Unit of measure  \
0  Percentage of female population age 15-19 with...              NaN   
1  Percentage of population age 15-19 with no edu...              NaN   
2  Percentage of female population age 15+ with n...              NaN   
3  Percentage of population age 15+ with no educa...              NaN   
4  Percentage of female population age 20-24 with...              NaN   

  Periodicity Base Period Other notes Aggregation method  \
0         NaN         NaN         NaN                NaN   
1         NaN         NaN         NaN                NaN   
2         NaN         NaN         NaN                NaN   
3         NaN         NaN         NaN                NaN   
4         NaN         NaN         NaN                NaN   

  Limitations and exceptions  Notes from original source General comments  \
0                        NaN                         NaN              NaN   
1                        NaN                         NaN              NaN   
2                        NaN                         NaN              NaN   
3                        NaN                         NaN              NaN   
4                        NaN                         NaN              NaN   

                                              Source  \
0  Robert J. Barro and Jong-Wha Lee: http://www.b...   
1  Robert J. Barro and Jong-Wha Lee: http://www.b...   
2  Robert J. Barro and Jong-Wha Lee: http://www.b...   
3  Robert J. Barro and Jong-Wha Lee: http://www.b...   
4  Robert J. Barro and Jong-Wha Lee: http://www.b...   

  Statistical concept and methodology Development relevance  \
0                                 NaN                   NaN   
1                                 NaN                   NaN   
2                                 NaN                   NaN   
3                                 NaN                   NaN   
4                                 NaN                   NaN   

  Related source links  Other web links  Related indicators  License Type  \
0                  NaN              NaN                 NaN           NaN   
1                  NaN              NaN                 NaN           NaN   
2                  NaN              NaN                 NaN           NaN   
3                  NaN              NaN                 NaN           NaN   
4                  NaN              NaN                 NaN           NaN   

   Unnamed: 20  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN

***
# Data Preprocessing
***

***
## 1. Indicators dataframe
***

***
## We will be interested in few indicators so our dataframe will be cleaner and will shrink
***

In [5]:
ed_ind = list()
for i in indicators_df['Topic']:
    if 'Education' in i:
        ed_ind.append(i)
ed_ind_arr = np.array(ed_ind)
ed_ind = list(np.unique(ed_ind_arr))

len(ed_ind)

4

In [6]:
fltr = (indicators_df['Topic'] == ed_ind[0]) | (indicators_df['Topic'] == ed_ind[1]) | (indicators_df['Topic'] == ed_ind[2]) | (indicators_df['Topic'] == ed_ind[3])
print(indicators_df.shape)
new_indicators_df = indicators_df[fltr]
print(new_indicators_df.shape)

(3665, 21)
(490, 21)


***
## We will also get rid of the details (most of the columns)
***

In [7]:
new_indicators_df = new_indicators_df[['Topic','Indicator Name']]

In [8]:
new_indicators_df

Topic                                     Indicator Name
360   Education Equality         DHS: Gross attendance rate. Post Secondary
361   Education Equality  DHS: Gross attendance rate. Post Secondary. Fe...
362   Education Equality   DHS: Gross attendance rate. Post Secondary. Male
363   Education Equality  DHS: Gross attendance rate. Post Secondary. Qu...
364   Education Equality  DHS: Gross attendance rate. Post Secondary. Qu...
...                  ...                                                ...
3462  Education Equality  UIS: Rate of out-of-school adolescents of lowe...
3463  Education Equality  UIS: Rate of out-of-school adolescents of lowe...
3464  Education Equality  UIS: Rate of out-of-school adolescents of lowe...
3465  Education Equality  UIS: Rate of out-of-school adolescents of lowe...
3466  Education Equality  UIS: Rate of out-of-school adolescents of lowe...

[490 rows x 2 columns]

In [9]:
new_indicators_df.isnull().sum()['Indicator Name']

0

In [10]:
new_indicators_df.shape[0] - len(new_indicators_df['Indicator Name'].unique())

0

***
## The indicators dataframe is clean and good to go 
***

***
## 2. Countries dataframe
***

In [11]:
countries_df = countries_df[['Country Code','Short Name','Currency Unit','Region']]

print(countries_df.isnull().sum())

countries_df.dropna(inplace=True)

countries_df

Country Code      0
Short Name        0
Currency Unit    26
Region           27
dtype: int64


Country Code    Short Name       Currency Unit                      Region
0            ABW         Aruba       Aruban florin   Latin America & Caribbean
1            AFG   Afghanistan      Afghan afghani                  South Asia
2            AGO        Angola      Angolan kwanza          Sub-Saharan Africa
3            ALB       Albania        Albanian lek       Europe & Central Asia
4            AND       Andorra                Euro       Europe & Central Asia
..           ...           ...                 ...                         ...
236          XKX        Kosovo                Euro       Europe & Central Asia
237          YEM         Yemen         Yemeni rial  Middle East & North Africa
238          ZAF  South Africa  South African rand          Sub-Saharan Africa
239          ZMB        Zambia  New Zambian kwacha          Sub-Saharan Africa
240          ZWE      Zimbabwe         U.S. dollar          Sub-Saharan Africa

[214 rows x 4 columns]

In [12]:
print(f"Number of rows : {countries_df.shape[0]}")
print('------------------------------------------------------------------')

for col in countries_df.columns:
    print(f"Number of unique values in the {col} column is {len(countries_df[col].unique())}")

Number of rows : 214
------------------------------------------------------------------
Number of unique values in the Country Code column is 214
Number of unique values in the Short Name column is 214
Number of unique values in the Currency Unit column is 151
Number of unique values in the Region column is 7


***
## Now the countries dataframe seems reasonable and good to go
***

***
## 3. Main dataframe
***

In [13]:
main_df = main_df.loc[:,:'2020']

tmp_new_indicators_df = new_indicators_df.set_index('Indicator Name')

new_main_df = pd.merge(main_df,tmp_new_indicators_df,left_on='Indicator Name',right_index=True)

new_main_df

Country Name Country Code  \
539                                      Arab World          ARB   
4204                            East Asia & Pacific          EAS   
7869    East Asia & Pacific (excluding high income)          EAP   
11534                                     Euro area          EMU   
15199                         Europe & Central Asia          ECS   
...                                             ...          ...   
871944                        Virgin Islands (U.S.)          VIR   
875609                           West Bank and Gaza          PSE   
879274                                  Yemen, Rep.          YEM   
882939                                       Zambia          ZMB   
886604                                     Zimbabwe          ZWE   

                                           Indicator Name    Indicator Code  \
539     DHS: Average years of schooling by age group. ...   HH.DHS.YRS.1519   
4204    DHS: Average years of schooling by age group. ...   HH.DHS.YRS.1519   
7869    DHS: Average years of schooling by age group. ...   HH.DHS.YRS.1519   
11534   DHS: Average years of schooling by age group. ...   HH.DHS.YRS.1519   
15199   DHS: Average years of schooling by age group. ...   HH.DHS.YRS.1519   
...                                                   ...               ...   
871944  UIS: Total net attendance rate, lower secondar...  UIS.NART.2.URB.M   
875609  UIS: Total net attendance rate, lower secondar...  UIS.NART.2.URB.M   
879274  UIS: Total net attendance rate, lower secondar...  UIS.NART.2.URB.M   
882939  UIS: Total net attendance rate, lower secondar...  UIS.NART.2.URB.M   
886604  UIS: Total net attendance rate, lower secondar...  UIS.NART.2.URB.M   

        1970  1971  1972  1973  1974  1975  1976  1977  1978  1979  1980  \
539      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4204     NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
7869     NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
11534    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
15199    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
...      ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
871944   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
875609   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
879274   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
882939   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
886604   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

        1981  1982  1983  1984  1985  1986  1987  1988  1989  1990  1991  \
539      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4204     NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
7869     NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
11534    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
15199    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
...      ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
871944   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
875609   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
879274   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
882939   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
886604   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

        1992  1993  1994  1995  1996  1997  1998  1999  2000       2001  2002  \
539      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN        NaN   NaN   
4204     NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN        NaN   NaN   
7869     NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN        NaN   NaN   
11534    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN        NaN   NaN   
15199    NaN   NaN   NaN   NaN   NaN   NaN

In [14]:
#Number of indicators in our dataset = number of rows
new_main_df.shape[0]

118580

***
## Of course we will not deal with more than 100 thousands indicators so we will try to go through the list of these indicators and select which could help us with our analysis and visualizations
***

In [15]:
for indicator in new_main_df['Indicator Name'].unique():
    print(indicator + '\n')

DHS: Average years of schooling by age group. Age 15-19

DHS: Average years of schooling by age group. Age 15-19. Female

DHS: Average years of schooling by age group. Age 15-19. Male

DHS: Average years of schooling by age group. Age 15-19. Quintile 1

DHS: Average years of schooling by age group. Age 15-19. Quintile 2

DHS: Average years of schooling by age group. Age 15-19. Quintile 3

DHS: Average years of schooling by age group. Age 15-19. Quintile 4

DHS: Average years of schooling by age group. Age 15-19. Quintile 5

DHS: Average years of schooling by age group. Age 15-19. Rural

DHS: Average years of schooling by age group. Age 15-19. Urban

DHS: Gini coefficient of average years of schooling. Age 15+

DHS: Gini coefficient of average years of schooling. Age 15+. Female

DHS: Gini coefficient of average years of schooling. Age 15+. Male

DHS: Gini coefficient of average years of schooling. Age 15+. Quintile 1

DHS: Gini coefficient of average years of schooling. Age 15+. Quinti

***
## At the first glance , it is obvious that we can choose our indicators whithout ignoring too much details , so these 118580 indicators will shrink remarkably and still , the dataset will provide us what we need for our analysis
***

# . . .